src/Strategy/Haskell/Cabal.hs

```diff
@@ -0,0 +1,180 @@
+{-# LANGUAGE TemplateHaskell #-}
+
+module Strategy.Haskell.Cabal
+  ( discover,
+
+    -- * Testing
+    BuildPlan (..),
+    PlanId (..),
+    InstallPlan (..),
+    PlanStyle (..),
+    PlanType (..),
+    buildGraph,
+  )
+where
+
+import Control.Effect.Diagnostics
+import Data.Aeson.Types
+import qualified Data.Map.Strict as M
+import qualified Data.Set as S
+import qualified Data.Text as T
+import Discovery.Walk
+import Effect.Exec
+import Effect.Grapher
+import Effect.ReadFS
+import qualified Graphing as G
+import Graphing (Graphing)
+import Prologue
+import Types
+
+newtype BuildPlan = BuildPlan {installPlans :: [InstallPlan]} deriving (Eq, Ord, Show)
+
+newtype PlanId = PlanId {unPlanId :: Text} deriving (FromJSON, Eq, Ord, Show)
+
+data InstallPlan
+  = InstallPlan
+      { planType :: PlanType,
+        planId :: PlanId,
+        planName :: Text,
+        planVersion :: Text,
+        planDepends :: Set PlanId,
+        planStyle :: Maybe PlanStyle,
+        planComponents :: Set PlanId
+      }
+  deriving (Eq, Ord, Show)
+
+data PlanStyle
+  = Local
+  | Global
+  deriving (Eq, Ord, Show)
+
+data PlanType
+  = PreExisting
+  | Configured
+  deriving (Eq, Ord, Show)
+
+instance FromJSON BuildPlan where
+  parseJSON = withObject "BuildPlan" $ \obj ->
+    BuildPlan <$> obj .: "install-plan"
+
+instance FromJSON InstallPlan where
+  parseJSON = withObject "InstallPlan" $ \obj ->
+    InstallPlan <$> (obj .: "type" >>= parsePlanType)
+      <*> obj .: "id"
+      <*> obj .: "pkg-name"
+      <*> obj .: "pkg-version"
+      <*> (fromMaybe mempty <$> obj .:? "depends")
+      <*> (obj .:? "style" >>= parsePlanStyle)
```

Comment:
this is good. we can improve it slightly with `traverse`:
```suggestion
      <*> (obj .:? "style" >>= traverse parsePlanStyle)
```

..where traverse has this type:
```hs
traverse :: (Text -> Parser PlanStyle) -> Maybe Text -> Parser (Maybe PlanStyle)
```
..which simplifies `parsePlanStyle`:

```hs
parsePlanStyle :: Text -> Parser PlanStyle
parsePlanStyle "global" = pure Global
parsePlanStyle "local" = pure Local
parsePlanStyle style = fail "unknown install plan style: " ++ T.unpack style
```

turns out `parsePlanStyle` has a more-general type, which GHC will tell you if you remove the type signature:
```hs
parsePlanStyle :: MonadFail m => Text -> m PlanStyle
```